src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/MSBuildShellOutNuGetProject.cs

```diff
@@ -0,0 +1,405 @@
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+using NuGet.Frameworks;
+using NuGet.LibraryModel;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.ProjectManagement;
+using NuGet.ProjectManagement.Projects;
+using NuGet.ProjectModel;
+using NuGet.Protocol.Core.Types;
+using EnvDTEProject = EnvDTE.Project;
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    public class MSBuildShellOutNuGetProject : NuGetProject, INuGetIntegratedProject
+    {
+        private const string MsBuild14Directory = @"C:\Program Files (x86)\MSBuild\14.0\Bin";
+        private const string MsBuild15Directory = @"C:\Program Files (x86)\MSBuild\15.0\Bin";
+        private const int MsBuildWaitTime = 2 * 60 * 1000; // 2 minutes in milliseconds
+
+        private readonly EnvDTEProject _envDTEProject;
+        private readonly IVsBuildPropertyStorage _buildPropertyStorage;
+
+        private readonly string _fullProjectPath;
+        private readonly string _projectName;
+        private readonly string _projectFullPath;
+        private readonly string _projectUniqueName;
+        private readonly string _msBuildDirectory;
+
+        private readonly object _installedPackagesLock = new object();
+        private List<PackageReference> _installedPackages;
+
+        public MSBuildShellOutNuGetProject(EnvDTEProject envDTEProject, IVsBuildPropertyStorage buildPropertyStorage)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            _envDTEProject = envDTEProject;
+            _buildPropertyStorage = buildPropertyStorage;
+
+            // Get information about the project from DTE.
+            _fullProjectPath = EnvDTEProjectUtility.GetFullProjectPath(_envDTEProject);
+            _projectName = _envDTEProject.Name;
+            _projectFullPath = EnvDTEProjectUtility.GetFullPath(_envDTEProject);
+            _projectUniqueName = EnvDTEProjectUtility.GetUniqueName(_envDTEProject);
+
+            // Detect the MSBuild directory based on the Visual Studio version number.
+            if (_envDTEProject.DTE.Version.StartsWith("15."))
+            {
+                _msBuildDirectory = MsBuild15Directory;
+            }
+            else
+            {
+                _msBuildDirectory = MsBuild14Directory;
+            }
+
+            // Add project metadata.
+            InternalMetadata.Add(NuGetProjectMetadataKeys.Name, _projectName);
+            InternalMetadata.Add(NuGetProjectMetadataKeys.FullPath, _projectFullPath);
+            InternalMetadata.Add(NuGetProjectMetadataKeys.UniqueName, _projectUniqueName);
+        }
+
+        public string BaseIntermediateOutputPath
+        {
+            get
+            {
+                var relativePath = ThreadHelper.JoinableTaskFactory.Run(async delegate
```

Comment:
Won't cache since this can change when the user modifies the project in VS. There is UI to change this value.